File: HW1/task3.py
from pyspark import SparkContext
import json
import sys
import time

sc= SparkContext("local[*]")
sc.setLogLevel("ERROR")

res = {}

input_file = sys.argv[1]
output_file = sys.argv[2]
type = sys.argv[3]
n_partitions = int(sys.argv[4])
n  = int(sys.argv[5])



if type == "customized":
    start=time.time()
    reviews = sc.textFile(sys.argv[1])
    reviews = reviews.map(lambda x: json.loads(x))
    #no_of_partitions = hash(id) as discussed in lecture
    def hashid(id):
        return hash(id)
    reviews = reviews.map(lambda x : (x['business_id'],1))
    reviews = reviews.partitionBy((n_partitions),hashid)
    reviews = reviews.reduceByKey(lambda x,y : x + y)
    res["n_partitions"] = reviews.getNumPartitions()
    items = reviews.mapPartitions(lambda it: [sum(1 for _ in it)])
    reviews = reviews.filter(lambda x: x[1] > n)
    
    res["n_items"] = items.collect()
    res["result"] = reviews.collect()
    process = time.time()-start
    print(process)
    
if type == "default":
    start=time.time()
#    time.sleep(0.5)
    reviews = sc.textFile(input_file)
    reviews = reviews.map(lambda x: json.loads(x))
    business = reviews.map(lambda x: (x['business_id'], 1))
    size =business.mapPartitions(lambda it: [sum(1 for _ in it)])
    business = business.reduceByKey(lambda x,y : x + y)
    res["n_partitions"] = business.getNumPartitions()
    #size = business.mapPartitions(lambda it: [sum(1 for _ in it)])
    business = business.filter(lambda x : x[1]>n)
    dup_bus={}
    i = 0
    for k,v in business.collect():
        dup_bus.update({k:v})
        i=i+1
        if i==n:
            dup_bus.update({"msg":"no_of_items"})
        elif i==n_partitions:
            dup_bus.update({"msg":"no_of_partitions"})
        elif i == i+n-n+i-10+10-50:
            dup_bus.update({"msg":"unwanted"})
        else:
            dup_bus.update({"msg":"helo"})
        
    
    res["n_items"] = size.collect()
    print(size.collect())
    res["result"] = business.collect()
    process_time = time.time() - start
    print(process_time)
    

    
with open(output_file,"w") as f:
    json.dump(res,f)
    

File: HW1/src/main/scala/task3.scala
import java.io.{File, PrintWriter}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession
import java.io.FileWriter
import org.apache.spark.HashPartitioner



object task3 {
  def main(args: Array[String]): Unit = {
    Logger.getLogger("org").setLevel(Level.ERROR)


    val spark = SparkSession.builder
      .master("local[*]")
      .appName("inf-553_task1")
      .getOrCreate()

    val input_file = args(0)
    val output_file = args(1)
    val typep = args(2)
    val n_partitions = args(3).toInt
    val n = args(4).toInt

    val sc = spark.sparkContext;
    if(typep=="default") {
      val data = spark.read.json(input_file).rdd
      val partitions = data.getNumPartitions
      val rdd = data.map(x => (x.getAs[String]("business_id"), 1)).reduceByKey((x, y) => x + y)
      val size = rdd.glom().map(_.length).collect().toList
      val filtered_rdd = rdd.filter(x => (x._2) > n).take(n).toList
      println("------------")
      println(partitions)
      println(size)
      println(filtered_rdd)

      val file = new FileWriter(output_file)
      file.write("{")
      file.write("\"n_partitions\": " + partitions + ",")
      file.write("\"n_items\": [")
      var j=1
      for (x <- size) {
        file.write("" + x)
        j=j+1
        print(j)
        if(j==partitions+1)
        {
          print(j)
          file.write("],")
        }
        else{
          print(j)
          file.write(",")
        }

      }

      file.write("\"result\": [")
      var i = 1
      for ((k, v) <- filtered_rdd) {
        file.write("[")
        file.write("\""+k+"\"")
        file.write("," + v)
        i = i+1
        if (i == filtered_rdd.length+1) {
          file.write("]")
        } else {
          file.write("],")
        }
      }
      file.write("]")
      file.write("}")
      file.flush()




    }

    else{
      val data = spark.read.json(input_file).rdd

      val rdd1 = data.map(x => (x.getAs[String]("business_id"), 1))
      val rdd2 =  rdd1.partitionBy(new HashPartitioner(n_partitions))
      val partitions = rdd2.getNumPartitions
      val size = rdd2.glom().map(_.length).collect().toList
      val new_rdd = rdd2.reduceByKey((x, y) => x + y).filter(x => (x._2) > n).take(n).toList
      println("---------")
      println(partitions)
      println(size)
      println(new_rdd)
      val file = new FileWriter(output_file)
      file.write("{")
      file.write("\"n_partitions\": " + partitions + ",")
      file.write("\"n_items\": [")
      var j=1
      for (x <- size) {

        file.write("" + x)
        j=j+1
        if(j==partitions+1)
        {
          file.write("],")
        }
        else{
          file.write(",")
        }

      }

      file.write("\"result\": [")
      var i = 1
      for ((k, v) <- new_rdd) {
        file.write("[")
        file.write("\""+k+"\"")
        file.write("," + v)
        i = i+1
        if (i == new_rdd.length+1) {
          file.write("]")
        } else {
          file.write("],")
        }
      }
      file.write("]")
      file.write("}")
      file.flush()

    }



//










  }


}
